Python Upgrade page (#437)
neo4j/driver-documentation: python-manual/modules/ROOT/pages/migration.adoc

= Upgrade & Migration

This page contains the list of new features and breaking changes that happened in the driver from version 4.4 to 5.LTS.
For a full list of changes, see the link:https://github.com/neo4j/neo4j-python-driver/wiki/5.x-changelog[Driver -> Changelog].

> The link:https://github.com/stefano-ottolenghi/neo4j-drivers-migration-assistent[Neo4j Drivers Migration Assistent] simplifies the upgrade process, by contextualizing the changelog to your codebase.

Reviewer: I assume this is only for the draft stage and the tool will be moved into some neo4j owned GitHub org.

Author: yes -- and not only that, maybe also be a pypi package.

The tool doesn't automatically rewrite your code; it only points at where action is needed, providing in-context information on how each hit should be addressed.

Any of the 5.x versions of the driver is compatible both with Neo4j server 4.4 and 5.x, so you can upgrade the driver before you upgrade the server.

[TIP]
When upgrading the Neo4j server to a newer version, the Cypher queries in your application may also need updating. +
See link:https://neo4j.com/docs/cypher-manual/current/deprecations-additions-removals-compatibility/[Cypher -> Deprecations, additions, and compatibility].

== Recommended steps

Using the migration assistent, the recommended upgrade steps are:

1. Run the migration assistent with its default configuration and address all the issues that it surfaces. The `path` argument supports globbing and can be repeated as many times as needed. +
+
[source, bash]
----
python main.py -l python <path...>
----
+
2. Re-run the assistent with the regex parser and check if any more legit issues are raised. +
+
[source, bash]
----
python main.py -l python --regex-parser <path...>
----
+

> The regex parser is able to parse implicit function calls and other hard-to-parse expressions, but is likely surface more false positive (for more information, see link:https://github.com/stefano-ottolenghi/neo4j-drivers-migration-assistent?tab=readme-ov-file#accuracy[Neo4j Drivers Migration Assistent -> Accuracy]).

Reviewer: ```suggestion
The regex parser is able to detect implicit function calls and other hard-to-parse expressions, but is likely to surface more false positives (for more information, see link:https://github.com/stefano-ottolenghi/neo4j-drivers-migration-assistent?tab=readme-ov-file#accuracy[Neo4j Drivers Migration Assistent -> Accuracy]).
```

You can add any spurious entry to the ignored list, so that the assistent won't bring them up in future runs.
3. Although the assistent can detect most of the locations where action is needed, a few changelog entries can't be surfaced in this form, so you should still revise the xref:breaking-changes[breaking changes] section and audit your codebase for the changes marked with {cross-mark}.
4. You are ready to upgrade the driver package to the new version. +
+
[source, bash]
----

> pip install -U neo4j

Reviewer: ```suggestion
pip install -U neo4j==5.*
```
What if 6 is out by the time someone follows this guide :innocent:

Reviewer: I guess we are assuming people are smart enough to transfer this to an entry in whatever format they use to track their dependencies (thinking of `requirements.txt`, `pyproject.toml`, etc.).

Author: > I guess we are assuming people are smart enough to transfer this to an entry in whatever format they use

Yes. I think the example still disambiguates any possible doubt around what "upgrading the package" may mean, and then people can map it to whatever system they use.

----

== New features

[cols="1a,1a"]
|===

|
.Support for Python 3.12
[%collapsible]
====

> The driver is compatible with any Python version starting from 3.6 up to 3.12.

Reviewer: ```suggestion
The driver is compatible with any Python version starting from 3.7 up to 3.12.
```

====
|
.`asyncio` support
[%collapsible]
====

> For using the driver's features in parallel via the link:https://docs.python.org/3/library/asyncio.html[asyncio] package.

Reviewer: ```suggestion
For using the driver's features asynchronously via the link:https://docs.python.org/3/library/asyncio.html[asyncio] package.
```
I first wanted to suggest "concurrently", because async in Python is *not* parallel (same in JS, but not in all languages). No threading will be involved, only one Task can run at any given time.
But I'm not sold on "concurrently" either, because you can have that in 4.4 as well using threads. So there's no real differentiation besides the "asyncio" that follows.

See xref:concurrency.adoc[].
====

|
.`Driver.execute_query()` to run transactions with less knowledge of driver's internals
[%collapsible]
====
The new function is a wrapper for `Session.execute_read/write()`, but it abstracts away the result processing part and returns a list of records to the caller directly.

See xref:query-simple.adoc[].
====
|
.Rust extension for performance
[%collapsible]
====

> The link:https://github.com/neo4j-drivers/neo4j-python-driver-rust-ext[Rust extension to the Python driver] is an alternative driver package that yields a 3x to 10x speedup compared to the regular driver.

Reviewer: ```suggestion
The link:https://github.com/neo4j-drivers/neo4j-python-driver-rust-ext[Rust extension to the Python driver] is an alternative driver package that yields an up to 10x speedup compared to the regular driver.
```
:innocent: maybe play it safe. I'm also in the process of updating the README as there are worst'pcase scenarios in which there is no significant difference in either direction.

> You can install it with `pip install neo4j-rust-ext`, either alongside the `neo4j` package or as a replacement to it.

Reviewer: ```suggestion
You can install it with `pip install neo4j-rust-ext`,  as a replacement to the `neo4j` package.
```
Written like that in the README for a purpose :) currently works to have both, might stop or become weird in the future.

Author: This falls in the conversation we had a few weeks ago on what the plans for the two packages are. 
If the two packages are going to stay separate long-term, and the rust extension is going to be supported long-term, then your change is good; but if in 1-2 years we decide to make the rust extension the default, then people may end up with a dangling dependency. 

Has the roadmap got clearer?

Author: Pending confirmation/implementation whether we make the rust extension the default.

Reviewer: Looks like we're doubling down on making it work the way you've documented it here.

See also https://github.com/neo4j/neo4j-python-driver-rust-ext/issues/20 and https://github.com/neo4j/neo4j-python-driver-rust-ext/pull/22/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R23-R24

See link:https://github.com/neo4j/neo4j-python-driver/discussions/921[Re-authentication].
====
|
.Mutual TLS (mTLS) as second authentication factor (2FA)
[%collapsible]
====
Allows for configuring client side TLS certificates to authenticate against the server.

See link:https://github.com/neo4j/neo4j-python-driver/discussions/1031[Mutual TLS].
====

|
.`BookmarkManager` support
[%collapsible]
====
Bookmark managers make it easier to achieve causal chaining of sessions.

See xref:bookmarks.adoc[].
====
|
.Notification filtering API
[%collapsible]
====
Filtering allows to receive only a subset of notifications from the server, and to improve performance server-side.

See xref:result-summary.adoc#_filter_notifications[Explore the query execution summary -> Filter notifications].
====

|
.GQL statuses of queries in the ResultSummary
[%collapsible]
====
Eventually, this API is planned to supersede the current notifications API.

See link:https://github.com/neo4j/neo4j-python-driver/discussions/1067[GQL statuses in ResultSummary].
====
|
.More convenient handling of server-side warnings and notifications
[%collapsible]
====

> You can configure the driver to emit warnings on notifications and/or to log them.

Reviewer: Not only that, by default, server-side warnings will be emitted as Python warning via `warning.warn(...)`. That might make a difference for people running in environments with warnings as errors.

Author: I copy-pasted this from the changelog without digging at all. I'll check.

See link:https://github.com/neo4j/neo4j-python-driver/discussions/1064[Convenient DBMS notifications].
====

|
.Telemetry
[%collapsible]
====
The driver sends anonymous API usage statistics to the server if requested.
Use the driver configuration `telemetry_disabled=True` to opt out.
====
|
.Concurrency misuse checks
[%collapsible]
====
For local development, run your application with `python -X dev ...` or set the environment variable `PYTHONNEO4JDEBUG` to anything non-empty to get additional concurrency misuse checks.
Some driver primitives (e.g. sessions) are not safe to be used concurrently, and using them so will lead to hard-to-debug errors.
This feature helps identifying such code paths (detection is still inherently racy, hence not guaranteed to always work).
====

|===

[[breaking-changes]]
== Breaking changes and deprecations

The column `Assistent` denotes whether the Neo4j Drivers Migration Assistent can detect issues related to the given changelog entry.
Entries with a cross mark {cross-mark} require manual inspection.

> Deprecated features get removed in the next LTS version.

Reviewer: That's not correct. Some of them will get removed in `6.0`. But possibly not all of them, we might decide to leave some deprecated APIs around for a prolonged migration period, especially if we anticipate much migration pain.

Author: "are likely to be removed in the next major release"?

|label:removed[Changed]
|{cross-mark}

|5.0
|link:https://neo4j.com/docs/api/python-driver/current/types/temporal.html#neo4j.time.Time[`Time`] -- The property `ticks_ns` has been renamed to `ticks`.
|label:removed[]
|{check-mark}

|5.0
|link:https://neo4j.com/docs/api/python-driver/current/types/temporal.html#neo4j.time.Time[`Time`] -- The property `hour_minute_second` has been removed. Use `hour_minute_second_nanosecond` instead.
|label:removed[]
|{check-mark}

|5.0
|Session methods `.read_transaction()` and `.write_transaction()` are deprecated in favor of link:https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Session.execute_read[`.execute_read()`] and link:https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Session.execute_write[`.execute_write()`].
Through the new methods, the first argument of transaction functions is a `ManagedTransaction` object. It behaves exactly like a regular `Transaction` object, except it does not offer the `.commit()`, `.rollback()`, `.close()`, and `.closed()` methods.
|label:deprecated[]
|{check-mark}

|5.3

> |Package-alias `neo4j-driver` is deprecated. It will stop receiving updates starting with 6.0.0. Install neo4j instead.

Reviewer: ```suggestion
|Package-alias `neo4j-driver` is deprecated. It will stop receiving updates starting with 6.0.0. Install `neo4j` instead.
```
